mito-ai/src/Extensions/InlineCompleter/provider.ts

```diff
@@ -0,0 +1,403 @@
+   */
+  async *stream(token: string): AsyncGenerator<{
+    response: IInlineCompletionItem;
+  }> {
+    if (!this.isEnabled()) {
+      throw new Error('Mito AI completion is disabled.');
+    }
+    // Wait for the acknowledge request to be fulfilled before starting the stream
+    await this._completionLock.promise;
+
+    if (this._currentToken !== token) {
+      // New completion may be triggered before the code have started streaming
+      // the previous one. So not raising an error here.
+      console.debug(`No stream found for token '${token}'.`);
+      return;
+    }
+
+    for await (const chunk of this._currentStream!) {
+      if (this._currentToken !== token) {
+        break;
+      }
+      yield chunk;
+      if (chunk.done || chunk.error) {
+        // Break this for loop
+        this._currentStream?.stop();
+      }
+    }
+  }
+
+  /**
+   * Extract prefix from request, accounting for context window limit.
+   */
+  private _getPrefix(request: CompletionHandler.IRequest): string {
+    return request.text.slice(0, request.offset);
+  }
+
+  /**
+   * Extract suffix from request, accounting for context window limit.
+   */
+  private _getSuffix(request: CompletionHandler.IRequest): string {
+    return request.text.slice(request.offset);
+  }
+
+  private _notifyCompletionFailure(error: CompletionError) {
+    Notification.emit(`Inline completion failed: ${error.type}`, 'error', {
+      autoClose: false,
+      actions: [
+        {
+          label: 'Show Traceback',
+          callback: () => {
+            showErrorMessage('Inline completion failed on the server side', {
+              message: error.traceback
+            });
+          }
+        }
+      ]
+    });
+  }
+
+  /**
+   * Process the stream chunk to make it available in the awaiting generator.
+   */
+  private _receiveStreamChunk(
+    _emitter: CompletionWebsocketClient,
+    chunk: InlineCompletionStreamChunk
+  ) {
+    if (chunk.error) {
+      this._notifyCompletionFailure(chunk.error);
+    }
+
+    const token = chunk.response.token;
+    if (!token) {
+      throw Error('Stream chunks must define `token` in `response`');
+    }
+
+    if (this._currentToken !== token) {
+      // This may happen if the backend is still streaming for a previous token
+      console.debug(
+        `Received completion chunk for an unknown token '${token}'`
+      );
+      return;
+    }
+
+    if (!this._currentStream) {
+      throw Error(`Stream not found for token ${token}`);
+    }
+
+    let fullCompletion = this._fullCompletionMap.get(this._currentStream) ?? '';
+    fullCompletion += chunk.response.insertText;
+    this._fullCompletionMap.set(this._currentStream, fullCompletion);
+
+    // Clean suggestion
+    let cleanedCompletion = fullCompletion.slice(0);
+    if (this._currentPrefix) {
+      if (
+        cleanedCompletion.startsWith(this._currentPrefix) ||
+        this._currentPrefix.startsWith(cleanedCompletion)
```

Comment:
the second case is due to the streaming of the completion. Taking back the first example:

`_currentPrefix -> "pri",`

The first completion chunk may be `"pr"` and so we don't want to display to the user `"pripr"`. Then when the second chunk arrives, like "int". The full completion will be `print` and we will have a proper cleaned completion `nt`.

Reply:
> Now that I think about this, maybe a better check to do is to check if we recalculated the currentPrefix, would it match exactly to the currentPrefix when the request was made.
> 
> This would account for the case where:
> 
>     1. The user continued typing
> 
>     2. The user deleted/changed the part of the prefix.
> 
> 
> Thoughts? I might be missing something.

Currently we cannot account easily for the new code changes by the user as any changes will discard the inline suggestion. But indeed I saw that copilot does that. Something that would require changing with JupyterLab core code. But certainly worth the improved user experienced.